File: python/src/wh40kdc/share.py
"""Compact, URL-safe list sharing — the Python mirror of the TS ``share`` module.

:func:`encode_share_token` packs a share list (the lossless essential subset of
a roster) into a ``share-v1`` token: registry-indexed unsigned-LEB128 varints,
base64url, no gzip. :func:`decode_share_token` reverses it against the embedded
registry. The wire format and registry are documented in
``tools/docs/share-token.md``; the ``conformance/share`` corpus pins this
implementation byte-for-byte against the TS and Rust ports.

Lists are plain dicts with the same camelCase keys the corpus uses, e.g.::

    {
        "name": "My List",
        "factionId": "adeptus-astartes",
        "detachmentIds": ["gladius-task-force"],
        "battleSize": "strike-force",
        "disposition": None,
        "units": [
            {
                "datasheetId": "intercessor-squad",
                "modelCount": 5,
                "isWarlord": True,
                "enhancementId": None,
                "allyFactionId": None,
                "allyRuleId": None,
                "attachedToOrdinal": None,
                "grants": [],
                "loadout": [["bolt-rifle", 5]],
            }
        ],
    }
"""

from __future__ import annotations

import base64
import json
from functools import lru_cache
from importlib import resources
from typing import Any

#: Current wire-format version. Bumped only if the byte layout changes.
SHARE_FORMAT_VERSION = 1

_FLAG_WARLORD = 1 << 0
_FLAG_ENH = 1 << 1
_FLAG_ATTACH = 1 << 2
_FLAG_ALLY = 1 << 3
_FLAG_GRANTS = 1 << 4

_BATTLE_SIZES = ("incursion", "strike-force")

SHARE_KINDS = (
    "faction",
    "detachment",
    "unit",
    "wargear",
    "enhancement",
    "ally_rule",
    "disposition",
)

ShareList = dict[str, Any]
DecodeResult = dict[str, Any]


class ShareEncodeError(ValueError):
    """An id absent from the registry — the only way encoding can fail."""


class _Malformed(Exception):
    """Internal: truncated or self-inconsistent byte stream."""


class _StaleRegistry(Exception):
    """Internal: a slot the embedded registry doesn't have."""


class _RegistryIndex:
    """Bidirectional id<->index lookup over one registry, with aliases folded in."""

    def __init__(self, registry: dict[str, Any]) -> None:
        self.version: int = registry["version"]
        kinds = registry.get("kinds", {})
        aliases: dict[str, str] = registry.get("aliases", {})
        self._to_index: dict[str, dict[str, int]] = {}
        self._from_index: dict[str, list[str]] = {}
        for kind in SHARE_KINDS:
            ids = kinds.get(kind, [])
            to_idx: dict[str, int] = {}
            frm: list[str] = []
            for i, id_ in enumerate(ids):
                to_idx[id_] = i
                # Decode resolves a slot to its current id (rewriting a rename).
                frm.append(aliases.get(id_, id_))
            # Encode must also find the current id at a renamed slot.
            for old_id, new_id in aliases.items():
                slot = to_idx.get(old_id)
                if slot is not None and new_id not in to_idx:
                    to_idx[new_id] = slot
            self._to_index[kind] = to_idx
            self._from_index[kind] = frm

    def index(self, kind: str, id_: str) -> int | None:
        return self._to_index[kind].get(id_)

    def id(self, kind: str, index: int) -> str | None:
        ids = self._from_index[kind]
        return ids[index] if 0 <= index < len(ids) else None


@lru_cache(maxsize=1)
def _embedded_index() -> _RegistryIndex:
    text = resources.files("wh40kdc").joinpath("_share_registry.json").read_text(encoding="utf-8")
    return _RegistryIndex(json.loads(text))


def share_registry_version() -> int:
    """Registry version this package embeds (stamped into every token it writes)."""
    return _embedded_index().version


# -- varint + base64url --------------------------------------------------------


def _write_varint(out: bytearray, value: int) -> None:
    if value < 0:
        raise ShareEncodeError(f"varint expects a non-negative integer, got {value}")
    v = value
    while v >= 0x80:
        out.append((v & 0x7F) | 0x80)
        v >>= 7
    out.append(v)


def _write_str(out: bytearray, s: str) -> None:
    encoded = s.encode("utf-8")
    _write_varint(out, len(encoded))
    out.extend(encoded)


class _Reader:
    def __init__(self, data: bytes) -> None:
        self._data = data
        self._pos = 0

    def byte(self) -> int:
        if self._pos >= len(self._data):
            raise _Malformed()
        b = self._data[self._pos]
        self._pos += 1
        return b

    def varint(self) -> int:
        result = 0
        shift = 0
        while True:
            b = self.byte()
            if shift >= 64:
                raise _Malformed()
            result |= (b & 0x7F) << shift
            if not (b & 0x80):
                return result
            shift += 7

    def string(self) -> str:
        length = self.varint()
        end = self._pos + length
        if end > len(self._data):
            raise _Malformed()
        chunk = self._data[self._pos : end]
        self._pos = end
        try:
            return chunk.decode("utf-8")
        except UnicodeDecodeError as e:  # pragma: no cover - defensive
            raise _Malformed() from e


def _b64url_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def _b64url_decode(token: str) -> bytes:
    pad = "=" * (-len(token) % 4)
    return base64.urlsafe_b64decode(token + pad)


# -- encode --------------------------------------------------------------------


def _require_index(idx: _RegistryIndex, kind: str, id_: str) -> int:
    i = idx.index(kind, id_)
    if i is None:
        raise ShareEncodeError(
            f'share registry has no {kind} id "{id_}" — run `npm run registry:build` '
            "and re-sync the bundles"
        )
    return i


def encode_share_list(list_: ShareList, registry: _RegistryIndex) -> str:
    """Encode a share list into a URL-safe ``share-v1`` token using ``registry``."""
    out = bytearray()
    out.append(SHARE_FORMAT_VERSION)
    _write_varint(out, registry.version)
    _write_str(out, list_.get("name", "") or "")

    faction_id = list_.get("factionId")
    _write_varint(out, 0 if faction_id is None else _require_index(registry, "faction", faction_id) + 1)
    battle = list_.get("battleSize", "strike-force")
    _write_varint(out, _BATTLE_SIZES.index(battle) if battle in _BATTLE_SIZES else 0)
    disposition = list_.get("disposition")
    _write_varint(
        out, 0 if disposition is None else _require_index(registry, "disposition", disposition) + 1
    )

    detachments = list_.get("detachmentIds", [])
    _write_varint(out, len(detachments))
    for det in detachments:
        _write_varint(out, _require_index(registry, "detachment", det))

    units = list_.get("units", [])
    _write_varint(out, len(units))
    for u in units:
        _write_varint(out, _require_index(registry, "unit", u["datasheetId"]))
        _write_varint(out, u["modelCount"])
        enhancement_id = u.get("enhancementId")
        attached = u.get("attachedToOrdinal")
        ally_faction = u.get("allyFactionId")
        ally_rule = u.get("allyRuleId")
        grants = u.get("grants") or []
        loadout = u.get("loadout") or []
        flags = (
            (_FLAG_WARLORD if u.get("isWarlord") else 0)
            | (_FLAG_ENH if enhancement_id else 0)
            | (_FLAG_ATTACH if attached is not None else 0)
            | (_FLAG_ALLY if (ally_rule or ally_faction) else 0)
            | (_FLAG_GRANTS if grants else 0)
        )
        _write_varint(out, flags)
        if enhancement_id:
            _write_varint(out, _require_index(registry, "enhancement", enhancement_id))
        if attached is not None:
            _write_varint(out, attached)
        if flags & _FLAG_ALLY:
            _write_varint(
                out, 0 if ally_faction is None else _require_index(registry, "faction", ally_faction) + 1
            )
            _write_varint(
                out, 0 if ally_rule is None else _require_index(registry, "ally_rule", ally_rule) + 1
            )
        if grants:
            _write_varint(out, len(grants))
            for g in grants:
                _write_str(out, g)
        _write_varint(out, len(loadout))
        for wid, count in loadout:
            _write_varint(out, _require_index(registry, "wargear", wid))
            _write_varint(out, count)

    return _b64url_encode(bytes(out))


def encode_share_token(list_: ShareList) -> str:
    """Encode a share list using the package's embedded registry."""
    return encode_share_list(list_, _embedded_index())


# -- decode --------------------------------------------------------------------


def _require_id(idx: _RegistryIndex, kind: str, slot: int) -> str:
    id_ = idx.id(kind, slot)
    if id_ is None:
        raise _StaleRegistry()
    return id_


def _decode_inner(token: str, registry: _RegistryIndex) -> ShareList:
    try:
        data = _b64url_decode(token)
    except Exception as e:  # invalid base64 → malformed
        raise _Malformed() from e
    r = _Reader(data)
    if r.byte() != SHARE_FORMAT_VERSION:
        raise _Malformed()
    r.varint()  # registry version — informational; bounds checks gate staleness

    name = r.string()
    faction_ref = r.varint()
    faction_id = None if faction_ref == 0 else _require_id(registry, "faction", faction_ref - 1)
    battle_idx = r.varint()
    battle_size = _BATTLE_SIZES[battle_idx] if battle_idx < len(_BATTLE_SIZES) else "strike-force"
    disposition_ref = r.varint()
    disposition = (
        None if disposition_ref == 0 else _require_id(registry, "disposition", disposition_ref - 1)
    )

    det_count = r.varint()
    detachment_ids = [_require_id(registry, "detachment", r.varint()) for _ in range(det_count)]

    unit_count = r.varint()
    units: list[dict[str, Any]] = []
    for _ in range(unit_count):
        datasheet_id = _require_id(registry, "unit", r.varint())
        model_count = r.varint()
        flags = r.varint()
        enhancement_id = (
            _require_id(registry, "enhancement", r.varint()) if flags & _FLAG_ENH else None
        )
        attached = r.varint() if flags & _FLAG_ATTACH else None
        ally_faction_id = None
        ally_rule_id = None
        if flags & _FLAG_ALLY:
            f_ref = r.varint()
            ally_faction_id = None if f_ref == 0 else _require_id(registry, "faction", f_ref - 1)
            rule_ref = r.varint()
            ally_rule_id = None if rule_ref == 0 else _require_id(registry, "ally_rule", rule_ref - 1)
        grants: list[str] = []
        if flags & _FLAG_GRANTS:
            for _ in range(r.varint()):
                grants.append(r.string())
        loadout: list[list[Any]] = []
        for _ in range(r.varint()):
            wid = _require_id(registry, "wargear", r.varint())
            loadout.append([wid, r.varint()])
        units.append(
            {
                "datasheetId": datasheet_id,
                "modelCount": model_count,
                "isWarlord": bool(flags & _FLAG_WARLORD),
                "enhancementId": enhancement_id,
                "allyFactionId": ally_faction_id,
                "allyRuleId": ally_rule_id,
                "attachedToOrdinal": attached,
                "grants": grants,
                "loadout": loadout,
            }
        )

    return {
        "name": name,
        "factionId": faction_id,
        "detachmentIds": detachment_ids,
        "battleSize": battle_size,
        "disposition": disposition,
        "units": units,
    }


def decode_share_list(token: str, registry: _RegistryIndex) -> DecodeResult:
    """Decode a ``share-v1`` token against ``registry``."""
    try:
        return {"ok": True, "list": _decode_inner(token, registry)}
    except _StaleRegistry:
        return {"ok": False, "reason": "stale-registry"}
    except Exception:
        return {"ok": False, "reason": "malformed"}


def decode_share_token(token: str) -> DecodeResult:
    """Decode a ``share-v1`` token using the package's embedded registry."""
    return decode_share_list(token, _embedded_index())

File: go/share.go
package wh40kdc

import (
	"encoding/base64"
	"encoding/json"
	"errors"
	"strings"
	"sync"
)

// Compact, URL-safe list sharing. encodeShareToken packs a share list into a
// share-v1 token (registry-indexed unsigned-LEB128 varints, base64url, no
// gzip); decodeShareToken reverses it against the embedded registry. Go mirror
// of python .../share.py; pinned by conformance/share byte-for-byte.

const shareFormatVersion = 1

const (
	flagWarlord = 1 << 0
	flagEnh     = 1 << 1
	flagAttach  = 1 << 2
	flagAlly    = 1 << 3
	flagGrants  = 1 << 4
)

var battleSizes = []string{"incursion", "strike-force"}

var shareKinds = []string{"faction", "detachment", "unit", "wargear", "enhancement", "ally_rule", "disposition"}

var errShareEncode = errors.New("share encode: id absent from registry")
var errMalformed = errors.New("malformed")
var errStaleRegistry = errors.New("stale-registry")

type registryIndex struct {
	version   int
	toIndex   map[string]map[string]int
	fromIndex map[string][]string
}

func newRegistryIndex(registry map[string]any) *registryIndex {
	ri := &registryIndex{
		version:   asInt(registry["version"]),
		toIndex:   map[string]map[string]int{},
		fromIndex: map[string][]string{},
	}
	kinds, _ := getMap(registry, "kinds")
	aliasesAny, _ := getMap(registry, "aliases")
	aliases := map[string]string{}
	for k, v := range aliasesAny {
		if s, ok := v.(string); ok {
			aliases[k] = s
		}
	}
	for _, kind := range shareKinds {
		ids := getStrList(kinds, kind)
		toIdx := map[string]int{}
		frm := make([]string, 0, len(ids))
		for i, id := range ids {
			toIdx[id] = i
			if alias, ok := aliases[id]; ok {
				frm = append(frm, alias)
			} else {
				frm = append(frm, id)
			}
		}
		for oldID, newID := range aliases {
			if slot, ok := toIdx[oldID]; ok {
				if _, exists := toIdx[newID]; !exists {
					toIdx[newID] = slot
				}
			}
		}
		ri.toIndex[kind] = toIdx
		ri.fromIndex[kind] = frm
	}
	return ri
}

func (ri *registryIndex) index(kind, id string) (int, bool) {
	i, ok := ri.toIndex[kind][id]
	return i, ok
}

func (ri *registryIndex) id(kind string, index int) (string, bool) {
	ids := ri.fromIndex[kind]
	if index < 0 || index >= len(ids) {
		return "", false
	}
	return ids[index], true
}

var (
	shareIndexOnce sync.Once
	shareIndexVal  *registryIndex
)

func embeddedShareIndex() *registryIndex {
	shareIndexOnce.Do(func() {
		var reg map[string]any
		if err := json.Unmarshal(shareRegistryJSON, &reg); err != nil {
			panic("wh40kdc: cannot parse embedded share_registry.json: " + err.Error())
		}
		shareIndexVal = newRegistryIndex(reg)
	})
	return shareIndexVal
}

// --- varint + base64url ---

func writeVarint(out *[]byte, value int) error {
	if value < 0 {
		return errShareEncode
	}
	v := uint64(value)
	for v >= 0x80 {
		*out = append(*out, byte((v&0x7f)|0x80))
		v >>= 7
	}
	*out = append(*out, byte(v))
	return nil
}

func writeStr(out *[]byte, s string) {
	b := []byte(s)
	_ = writeVarint(out, len(b))
	*out = append(*out, b...)
}

type reader struct {
	data []byte
	pos  int
}

func (r *reader) byte() (byte, error) {
	if r.pos >= len(r.data) {
		return 0, errMalformed
	}
	b := r.data[r.pos]
	r.pos++
	return b, nil
}

func (r *reader) varint() (int, error) {
	result := 0
	shift := 0
	for {
		b, err := r.byte()
		if err != nil {
			return 0, err
		}
		if shift >= 64 {
			return 0, errMalformed
		}
		result |= int(b&0x7f) << shift
		if b&0x80 == 0 {
			return result, nil
		}
		shift += 7
	}
}

func (r *reader) string() (string, error) {
	length, err := r.varint()
	if err != nil {
		return "", err
	}
	end := r.pos + length
	if end > len(r.data) || length < 0 {
		return "", errMalformed
	}
	chunk := r.data[r.pos:end]
	r.pos = end
	return string(chunk), nil
}

// --- encode ---

func requireIndex(ri *registryIndex, kind, id string) (int, error) {
	i, ok := ri.index(kind, id)
	if !ok {
		return 0, errShareEncode
	}
	return i, nil
}

func encodeShareList(list map[string]any, ri *registryIndex) (string, error) {
	var out []byte
	out = append(out, shareFormatVersion)
	if err := writeVarint(&out, ri.version); err != nil {
		return "", err
	}
	writeStr(&out, strOr(list, "name", ""))

	if list["factionId"] == nil {
		_ = writeVarint(&out, 0)
	} else {
		i, err := requireIndex(ri, "faction", getStr(list, "factionId"))
		if err != nil {
			return "", err
		}
		_ = writeVarint(&out, i+1)
	}
	battle := strOr(list, "battleSize", "strike-force")
	bi := 0
	for idx, b := range battleSizes {
		if b == battle {
			bi = idx
		}
	}
	_ = writeVarint(&out, bi)
	if list["disposition"] == nil {
		_ = writeVarint(&out, 0)
	} else {
		i, err := requireIndex(ri, "disposition", getStr(list, "disposition"))
		if err != nil {
			return "", err
		}
		_ = writeVarint(&out, i+1)
	}

	detachments := getStrList(list, "detachmentIds")
	_ = writeVarint(&out, len(detachments))
	for _, det := range detachments {
		i, err := requireIndex(ri, "detachment", det)
		if err != nil {
			return "", err
		}
		_ = writeVarint(&out, i)
	}

	units := getList(list, "units")
	_ = writeVarint(&out, len(units))
	for _, uAny := range units {
		u, _ := asMap(uAny)
		i, err := requireIndex(ri, "unit", getStr(u, "datasheetId"))
		if err != nil {
			return "", err
		}
		_ = writeVarint(&out, i)
		_ = writeVarint(&out, asInt(u["modelCount"]))
		enhancementID := getStr(u, "enhancementId")
		attached := u["attachedToOrdinal"]
		allyFaction := getStr(u, "allyFactionId")
		allyRule := getStr(u, "allyRuleId")
		grants := getStrList(u, "grants")
		loadout := getList(u, "loadout")
		flags := 0
		if truthy(u["isWarlord"]) {
			flags |= flagWarlord
		}
		if enhancementID != "" {
			flags |= flagEnh
		}
		if attached != nil {
			flags |= flagAttach
		}
		if allyRule != "" || allyFaction != "" {
			flags |= flagAlly
		}
		if len(grants) > 0 {
			flags |= flagGrants
		}
		_ = writeVarint(&out, flags)
		if enhancementID != "" {
			i, err := requireIndex(ri, "enhancement", enhancementID)
			if err != nil {
				return "", err
			}
			_ = writeVarint(&out, i)
		}
		if attached != nil {
			_ = writeVarint(&out, asInt(attached))
		}
		if flags&flagAlly != 0 {
			if allyFaction == "" {
				_ = writeVarint(&out, 0)
			} else {
				i, err := requireIndex(ri, "faction", allyFaction)
				if err != nil {
					return "", err
				}
				_ = writeVarint(&out, i+1)
			}
			if allyRule == "" {
				_ = writeVarint(&out, 0)
			} else {
				i, err := requireIndex(ri, "ally_rule", allyRule)
				if err != nil {
					return "", err
				}
				_ = writeVarint(&out, i+1)
			}
		}
		if len(grants) > 0 {
			_ = writeVarint(&out, len(grants))
			for _, g := range grants {
				writeStr(&out, g)
			}
		}
		_ = writeVarint(&out, len(loadout))
		for _, pairAny := range loadout {
			pair, _ := asList(pairAny)
			if len(pair) != 2 {
				return "", errShareEncode
			}
			wid, _ := pair[0].(string)
			i, err := requireIndex(ri, "wargear", wid)
			if err != nil {
				return "", err
			}
			_ = writeVarint(&out, i)
			_ = writeVarint(&out, asInt(pair[1]))
		}
	}
	return base64.RawURLEncoding.EncodeToString(out), nil
}

func encodeShareToken(list map[string]any) (string, error) {
	return encodeShareList(list, embeddedShareIndex())
}

// --- decode ---

func requireID(ri *registryIndex, kind string, slot int) (string, error) {
	id, ok := ri.id(kind, slot)
	if !ok {
		return "", errStaleRegistry
	}
	return id, nil
}

func b64urlDecode(token string) ([]byte, error) {
	pad := strings.Repeat("=", (4-len(token)%4)%4)
	return base64.URLEncoding.DecodeString(token + pad)
}

func decodeInner(token string, ri *registryIndex) (map[string]any, error) {
	data, err := b64urlDecode(token)
	if err != nil {
		return nil, errMalformed
	}
	r := &reader{data: data}
	b, err := r.byte()
	if err != nil || b != shareFormatVersion {
		return nil, errMalformed
	}
	if _, err := r.varint(); err != nil { // registry version — informational
		return nil, err
	}
	name, err := r.string()
	if err != nil {
		return nil, err
	}
	factionRef, err := r.varint()
	if err != nil {
		return nil, err
	}
	var factionID any
	if factionRef != 0 {
		id, err := requireID(ri, "faction", factionRef-1)
		if err != nil {
			return nil, err
		}
		factionID = id
	}
	battleIdx, err := r.varint()
	if err != nil {
		return nil, err
	}
	battleSize := "strike-force"
	if battleIdx >= 0 && battleIdx < len(battleSizes) {
		battleSize = battleSizes[battleIdx]
	}
	dispositionRef, err := r.varint()
	if err != nil {
		return nil, err
	}
	var disposition any
	if dispositionRef != 0 {
		id, err := requireID(ri, "disposition", dispositionRef-1)
		if err != nil {
			return nil, err
		}
		disposition = id
	}
	detCount, err := r.varint()
	if err != nil {
		return nil, err
	}
	detachmentIDs := []any{}
	for i := 0; i < detCount; i++ {
		slot, err := r.varint()
		if err != nil {
			return nil, err
		}
		id, err := requireID(ri, "detachment", slot)
		if err != nil {
			return nil, err
		}
		detachmentIDs = append(detachmentIDs, id)
	}
	unitCount, err := r.varint()
	if err != nil {
		return nil, err
	}
	units := []any{}
	for i := 0; i < unitCount; i++ {
		slot, err := r.varint()
		if err != nil {
			return nil, err
		}
		datasheetID, err := requireID(ri, "unit", slot)
		if err != nil {
			return nil, err
		}
		modelCount, err := r.varint()
		if err != nil {
			return nil, err
		}
		flags, err := r.varint()
		if err != nil {
			return nil, err
		}
		var enhancementID any
		if flags&flagEnh != 0 {
			s, err := r.varint()
			if err != nil {
				return nil, err
			}
			id, err := requireID(ri, "enhancement", s)
			if err != nil {
				return nil, err
			}
			enhancementID = id
		}
		var attached any
		if flags&flagAttach != 0 {
			v, err := r.varint()
			if err != nil {
				return nil, err
			}
			attached = float64(v)
		}
		var allyFactionID, allyRuleID any
		if flags&flagAlly != 0 {
			fRef, err := r.varint()
			if err != nil {
				return nil, err
			}
			if fRef != 0 {
				id, err := requireID(ri, "faction", fRef-1)
				if err != nil {
					return nil, err
				}
				allyFactionID = id
			}
			ruleRef, err := r.varint()
			if err != nil {
				return nil, err
			}
			if ruleRef != 0 {
				id, err := requireID(ri, "ally_rule", ruleRef-1)
				if err != nil {
					return nil, err
				}
				allyRuleID = id
			}
		}
		grants := []any{}
		if flags&flagGrants != 0 {
			n, err := r.varint()
			if err != nil {
				return nil, err
			}
			for j := 0; j < n; j++ {
				g, err := r.string()
				if err != nil {
					return nil, err
				}
				grants = append(grants, g)
			}
		}
		loadout := []any{}
		nL, err := r.varint()
		if err != nil {
			return nil, err
		}
		for j := 0; j < nL; j++ {
			s, err := r.varint()
			if err != nil {
				return nil, err
			}
			wid, err := requireID(ri, "wargear", s)
			if err != nil {
				return nil, err
			}
			count, err := r.varint()
			if err != nil {
				return nil, err
			}
			loadout = append(loadout, []any{wid, float64(count)})
		}
		units = append(units, map[string]any{
			"datasheetId":       datasheetID,
			"modelCount":        float64(modelCount),
			"isWarlord":         flags&flagWarlord != 0,
			"enhancementId":     enhancementID,
			"allyFactionId":     allyFactionID,
			"allyRuleId":        allyRuleID,
			"attachedToOrdinal": attached,
			"grants":            grants,
			"loadout":           loadout,
		})
	}
	return map[string]any{
		"name":          name,
		"factionId":     factionID,
		"detachmentIds": detachmentIDs,
		"battleSize":    battleSize,
		"disposition":   disposition,
		"units":         units,
	}, nil
}

func decodeShareList(token string, ri *registryIndex) map[string]any {
	list, err := decodeInner(token, ri)
	if err != nil {
		if errors.Is(err, errStaleRegistry) {
			return map[string]any{"ok": false, "reason": "stale-registry"}
		}
		return map[string]any{"ok": false, "reason": "malformed"}
	}
	return map[string]any{"ok": true, "list": list}
}

func decodeShareToken(token string) map[string]any {
	return decodeShareList(token, embeddedShareIndex())
}
